doc/man/man4/s_cdi.4.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man4/s_cdi.4.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH S_CDI 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Devices"
.SH NAME
.B s_cdi
\- SpiderWAN to CDI Conversion Module
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <uinstd.h>
.br
.B #include <stropts.h>
.HP 12
\fIwan_stream\fP = \fBopen\fP(\fIwan_driver\fP, O_RDWR);
.HP
\fIretval\fP = \fBioctl\fP(\fIwan_stream\fP, I_PUSH, \*(lq\fBs_cdi\fP\*(rq);
.\"
.\"
.SH DESCRIPTION
.PP
.BR "@PACKAGE_TITLE@"
and most OpenSS7 packages, utilize the Communications Device Interface (CDI),
.[
cdi revision @PACKAGE_VERSION@ draft @PACKAGE_RELEASE@
.]
.BR cdi (7),
as an interface to WAN devices.  However, due to the widespread use in the
1990's of systems deploying SpiderWAN, SpiderX.25
.[
spiderx25wan
.]
and SpiderISDN
.[
spiderisdnwan
.]
a number of applications programs, as well as \fISTREAMS\fP(9) drivers and
modules may have been written to provide these interfaces.  Therefore, to
provide compatibility between the OpenSS7 CDI implementations of applications
and user modules or drivers expecting a CDI interface, the \fBS_CDI\fP module
was developed to provide source compatibility with these modules or drivers.
.PP
The \fBS_CDI\fP module is a pushable \fISTREAMS\fP(9) module that performs
conversion between the Spider
.IR "Wide Area Network (WAN)" ,
.[
spiderx25wan
.]
interface at the lower service interface, and the
.IR "Communications Device Interface (CDI)" ,
.[
cdi revision @PACKAGE_VERSION@ draft @PACKAGE_RELEASE@
.]
.BR cdi (7),
at the upper service interface.
.\"
.\"
.SH USAGE
.PP
.B S_CDI
is not normally pushed by user applications.  Rather, it is pushed as necessary
by OpenSS7 stack configuration programs or as part of an extended interface for
a WAN device driver using an
.BR autopush (8)
specification in configuration files, or an
.BR autopush (9)
specification within the device driver itself.
.\"
.\"
.SH PRIMITIVES
.PP
.B S_CDI
accepts and issues Communications Device Interface (CDI),
.[
cdi revision @PACKAGE_VERSION@ draft @PACKAGE_RELEASE@
.]
.BR cdi (7),
service primitives at its upper service interface boundary.
.B S_CDI
accepts and issues Wide Area Network (WAN) Interface,
.[
WAN Revision @PACKAGE_VERSION@ Draft @PACKAGE_RELEASE@
.]
.BR wan (7),
services primitives at its lower service interface boundary.
.\"
.\"
.SH IOCTLS
.PP
While the underlying driver should support the WAN input-output controls
described under
.BR wan (7),
the
.B S_CDI
module also supports CDI input-output controls described for the OpenSS7
Project drivers under
.BR cdi (7).
.\"
.\"
.SH NOTICES
.PP
The Spider Wide Area Network (WAN) interface is a proprietary interface without
a \fIformal\fP standard (although its widespread use makes it an \fIipso
facto\fP standard).  Either the Communications Device Interface (CDI),
.[
cdi revision @PACKAGE_VERSION@ draft @PACKAGE_RELEASE@
.]
.BR cdi (7),
or the Data Link Provider Interface (DLPI),
.[
dlpi revision 2.0.0 draft 2
.]
.BR dlpi (7),
should be used in its stead.
.\"
.\"
.SH MODULES
.PP
.TP
.B s_cdi
The pushable
.IR STREAMS (9)
module named \*(lq\fBs_cdi\fP\*(rq.
.\"
.\"
.SH BUGS
.PP
.B S_CDI
has no known bugs.
.\"
.\"
.SH SEE ALSO
.PP
.BR cdi (7),
.BR autopush (8),
.IR STREAMS (9).
.\"
.\"
.SH COMPATIBILITY
.PP
.B S_CDI
is compatible with the Spider WAN interface,
.[
WAN Revision @PACKAGE_VERSION@ Draft @PACKAGE_RELEASE@
.]
and implementations based on 
.IR SpiderWAN ,
such as
.IR "AIXlink/X.25" ,
.IR "HP-UX X.25" ,
.IR "IRIS SX.25" ,
.IR "Solstice X.25" ,
.IR "PT X.25" , 
.IR "SBE X.25" , 
.IR "RadiSys WAN" ,
and others,
with the following compatibility considerations:
.IP \(em \w'\(em\(em'u
Although the
.BR wan (7)
interface is an \fiipso facto\fP standard, it should not be used by portable
programs.  The Open Group Data Link Provider Interface (DLPI),
.[
dlpi revision 2.0.0 draft 2
.]
.BR dlpi (7),
should be used in its stead.
.IP \(em
.B @PACKAGE_TITLE@
attempts to maintain
.I source compatibility
with the core features provided by most implementations, but compatibility with
all features of all implementations is not possible.  Your mileage may vary.
.RS
.PP
.B @PACKAGE_TITLE@
provides a core
.I source compatible
implementation of the Spider WAN interface as described in OpenSS7 project
documentation,
.[
WAN Revision @PACKAGE_VERSION@ Draft @PACKAGE_RELEASE@
.]
for the purpose of porting legacy WAN applications, drivers and modules to
.B Linux
with a minimal porting effort.  While
.I source compatibility
with core features is attempted,
.I binary compatibility
is not attempted.  Any binary compatibility experienced could be broken in a
subsequent release.
.PP
New applications should use the Communications Device Interface (CDI),
.[
cdi revision @PACKAGE_VERSION@ draft @PACKAGE_RELEASE@
.]
.BR cdi (7),
instead.
.RE
.PP
For additional compatibility information, see
.BR dlpi (7),
.BR cdi (7),
.BR wan (7),
.IR STREAMS (9).
.\"
.\"
.SH CONFORMANCE
.PP
CDI,
.[
cdi revision @PACKAGE_VERSION@ draft @PACKAGE_RELEASE@
.]
WAN.
.[
wan revision @PACKAGE_VERSION@ draft @PACKAGE_RELEASE@
.]
.\"
.\"
.SH HISTORY
.PP
The
.B S_CDI
module first appears in OpenSS7
.BR strx25-0.9.2.1 .
.\"
.\"
.SH VERSIONS
.PP
This manual page was written for the
.B @PACKAGE_TITLE@
package
.RB ( "@PACKAGE@-@VERSION@" ).
.PP
This version of the
.B S_CDI
module was developed for the OpenSS7 Project, and included in the
.BR "@PACKAGE_TITLE@" .
.\"
.\"
.[
$LIST$
.]
.TI


NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


src/modules/s_cdi.c
/*****************************************************************************

 @(#) File: src/modules/s_cdi.c

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 *****************************************************************************/

static char const ident[] = "src/modules/s_cdi.c (" PACKAGE_ENVR ") " PACKAGE_DATE;

/*
 * S_CDI is a simple conversion module that converts between a CDI Stream
 * supporting a WAN interface and the SpiderWAN raw HDLC and ASYNC interface.
 * This is simply to provide a conversion module for this utilities or legacy
 * application that need it.  This module may be autopushed onto a clone or
 * non-clone minor device number of a device driver Stream.
 */

#define _SVR4_SOURCE
#define _MPS_SOURCE
#define _SUN_SOURCE

#include <sys/os7/compat.h>

#define S_CDI_DESCRIP	"S_CDI STREAMS Module"
#define S_CDI_EXTRA	"Part of the OpenSS7 SS7 Stack for Linux Fast-STREAMS"
#define S_CDI_REVISION	"OpenSS7 src/modules/s_cdi.c (" PACKAGE_ENVR ") " PACKAGE_DATE
#define S_CDI_COPYRIGHT	"Copyright (c) 2008-2015  Monavacon Limited.  All Rights Reserved."
#define S_CDI_DEVICE	"Provides OpenSS7 SpiderWAN to CDI (S_CDI)"
#define S_CDI_CONTACT	"Brian Bidulock <bidulock@openss7.org>"
#define S_CDI_LICENSE	"GPL"
#define S_CDI_BANNER	S_CDI_DESCRIP	"\n" \
			S_CDI_EXTRA	"\n" \
			S_CDI_REVISION	"\n" \
			S_CDI_COPYRIGHT	"\n" \
			S_CDI_DEVICE	"\n" \
			S_CDI_CONTACT	"\n"
#define S_CDI_SPLASH	S_CDI_DEVICE	" - " \
			S_CDI_REVISION	"\n"

#ifdef LINUX
MODULE_AUTHOR(S_CDI_CONTACT);
MODULE_DESCRIPTION(S_CDI_DESCRIP);
MODULE_SUPPORTED_DEVICE(S_CDI_DEVICE);
#ifdef MODULE_LICENSE
MODULE_LICENSE(S_CDI_LICENSE);
#endif				/* MODULE_LICENSE */
#ifdef MODULE_ALIAS
MODULE_ALIAS("streams-s_cdi");
#endif				/* MODULE_ALIAS */
#ifdef MODULE_VERSION
MODULE_VERSION(PACKAGE_ENVR);
#endif				/* MODULE_VERSION */
#endif				/* LINUX */

#define S_CDI_MOD_ID	CONFIG_STREAMS_S_CDI_MODID
#define S_CDI_MOD_NAME	CONFIG_STREAMS_S_CDI_NAME

#ifndef S_CDI_MOD_NAME
#define S_CDI_MOD_NAME	"s_cdi"
#endif				/* S_CDI_MOD_NAME */

#ifndef S_CDI_MOD_ID
#define S_CDI_MOD_ID	0
#endif				/* S_CDI_MOD_ID */

/*
 *  STREAMS DEFINITIONS
 *  -------------------------------------------------------------------------
 */

#define MOD_ID		S_CDI_MOD_ID
#define MOD_NAME	S_CDI_MOD_NAME
#ifdef MODULE
#define MOD_BANNER	S_CDI_BANNER
#else				/* MODULE */
#define MOD_BANNER	S_CDI_SPLASH
#endif				/* MODULE */

/*
 *  LINUX INITIALIZATION
 *  -------------------------------------------------------------------------
 */

#ifdef LINUX
unsigned short modid = MOD_ID;

#ifndef module_param
MODULE_PARM(modid, "h");
#else				/* module_param */
module_param(modid, short, 0444);
#endif				/* module_param */
MODULE_PARM_DESC(modid, "Module ID for S_CDI module.  (0 for allocation.)");

/** s_cdiinit - initialize S_CDI
  */
static __init int
s_cdiinit(void)
{
	cmn_err(CE_NOTE, MOD_BANNER);
	return (0);
}

/** s_cdiexit - terminate S_CDI
  */
static __exit void
s_cdiexit(void)
{
	return;
}

module_init(s_cdiinit);
module_exit(s_cdiexit);

#endif				/* LINUX */
